[WIP][Button] Add pressed state to

AndrewMusgrave opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [ComponentName] (if applicable), for example: [Button]
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Prefix it with [WIP] while it’s a work in progress
-->

### WHY are these changes introduced?

Fixes #0000 <!-- link to issue if one exists -->

<!--
  Context about the problem that’s being addressed.
-->

### WHAT is this pull request doing?

<!--
  Summary of the changes committed.

  Before / after screenshots are appreciated for UI changes. Make sure to include alt text that describes the screenshot.

  If you include an animated gif showing your change, wrapping it in a details tag is recommended. Gifs usually autoplay, which can cause accessibility issues for people reviewing your PR:

    <details>
      <summary>Summary of your gif(s)</summary>
      <img src="..." alt="Description of what the gif shows">
    </details>
-->

## <!-- ℹ️ Delete the following for small / trivial changes -->

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<!--
  Give as much information as needed to experiment with the component
  in the playground.
-->

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}

```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the [Polaris UI kit](https://polaris.shopify.com/resources/polaris-ui-kit)

<!--
  When tophatting, please check the UNRELEASED entry for consistency, as per the [guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md).
-->

<!--
  If you don't have access to Percy, please request access in the #polaris Slack channel.
-->

AndrewMusgrave commented:
Slight delay - Waiting to pair with a designer to hammer out visual states

AndrewMusgrave commented on src/components/Button/Button.tsx:
```diff
@@ -61,7 +63,10 @@ export interface ButtonProps {
   ariaExpanded?: boolean;
-  /** Tells screen reader the element is pressed */
+  /**
+   * @deprecated As of release 4.X.X, replaced by {@link https://polaris.shopify.com/components/structure/page#props-pressed}
```
TODO: Edit this while merging

AndrewMusgrave commented on src/styles/shared/_buttons.scss:
```diff
@@ -129,8 +129,46 @@
-@mixin button-outline($outline-color) {
-  background: transparent;
+@mixin button-filled-pressed($button-color, $focus-color) {
+  background: linear-gradient(
```
Grabbed styles from ToggleButton in web

tmlayton commented:
Hey @AndrewMusgrave, can we hold off on this until after we make the changes with CSS custom properties and theming in the next few months? Otherwise, let’s pair on reworking this mixin to avoid color manipulation as it makes the code for backwards compatibility kind of gross. Thanks!

BPScott commented:
I got tagged as this is modifying the public sass API. Eventually we want to stop touching that but we've not got the time to bash out that plan. The change to that looks fine for now.

Removing myself from review as I haven't got the time to look at the rest of this

AndrewMusgrave commented:
> Hey AndrewMusgrave, can we hold off on this until after we make the changes with CSS custom properties and theming in the next few months? Otherwise, let’s pair on reworking this mixin to avoid color manipulation as it makes the code for backwards compatibility kind of gross. Thanks! 


A few months is a little bit longer to let this go stale, I'll remove all the sass manipulation 😄 

> Can you add an example so folks can easily tophat this change and see it in use in our documentation once it's released?

Good call, I'll include an example. There are a lot of button variants though. The PR description has a complex playground to see the pressed states for 🎩 

@mirualves Want to spend a few minutes going over the visuals Monday for a final check?

danrosenthal commented on src/components/Button/Button.scss:
```diff
@@ -7,6 +7,17 @@ $slim-vertical-padding: ($slim-min-height - line-height(body) - rem(2px)) / 2;
+$destructive-pressed-lightest: rgb(176, 43, 19);
+$destructive-pressed-light: rgb(167, 41, 18);
+$destructive-pressed-dark: rgb(117, 4, 10);
+$destructive-pressed-darkest: rgb(93, 3, 8);
```
I've got some concerns on the impact all these snowflake colors will have on the color system work, and the move to client-side theming. Some context on how these colors were generated would give us a stronger ability to upgrade this to consume the color system. I actually think this was better with the `darken` and `rgba` functions, since we know where the colors started and how they got to the final state. That feels easier to systematically reproduce.

@tmlayton, what do you think about going back to the way it was before, given what I said above?

danrosenthal replied:
Either that, or leave comments on each of these colors to show how they were achieved using programmatic darkening of our color tokens.

dleroux replied:
I agree with Dan here. If there were achieved using the `darken` functions it might be best to leave it. It does make it clear on how to reproduce across themes.

AndrewMusgrave replied:
Added comments on how the colors were achieved

AndrewMusgrave commented:
> Make sure there's a test passing the pressed property and this should be good to go.

I added a few tests. Historically we haven't tested visuals, is this something we're starting to do?

@danrosenthal I moved the new mixings into `Button.scss` so they're not exposed

danrosenthal commented:
> I added a few tests. Historically we haven't tested visuals, is this something we're starting to do?

It's not so much about testing visuals, as it is just having coverage on all code paths. I'd even be fine with adding `pressed` to an existing test just to get that code path to execute. The presence of the classNames is less interesting, and can be covered by VRT.

> @danrosenthal I moved the new mixings into Button.scss so they're not exposed

Fantastic!

dleroux commented on src/components/Button/Button.tsx:
```diff
@@ -39,6 +39,8 @@ export interface ButtonProps {
   textAlign?: TextAlign;
   /** Gives the button a subtle alternative to the default button styling, appropriate for certain backdrops */
   outline?: boolean;
+  /** Demonstrates a pressed state */
```
```suggestion
  /** Gives the button the appearance of being pressed */
```
